src/bootupd.rs

```diff
@@ -489,6 +496,88 @@ pub(crate) fn client_run_validate() -> Result<()> {
+    if bootloader == "none" {
+        println!("ostree repo 'sysroot.bootloader' config option already set to 'none'.");
+        println!("Assuming that the migration is already complete.");
+        File::create(stamp_file)?;
+        return Ok(());
+    }
+
+    // Remount /boot read write just for this unit (we are called in a slave mount namespace by systemd)
+    ensure_writable_boot();
+
+    let grub_config_dir = PathBuf::from("/boot/grub2");
+    let dirfd = openat::Dir::open(grub_config_dir).ok_or_else(|| anyhow!("Could not open {}. Maybe /boot is not mounted? Not performing migration.", grub_config_dir));
+
+    // Migrate /boot/grub2/grub.cfg to a static GRUB config if it is a symlink
+    let grub_config_filename ="/boot/grub2/grub.cfg";
+    match dirfd.read_link("grub.cfg") {
+        Err(_) => {
+            println!("/boot/grub2/grub.cfg is not a symlink. Nothing to migrate.");
+        },
+        Ok(path) => {
+            println!("Migrating to a static GRUB config...");
+
+            // Backup the current GRUB config which is hopefully working right now
+            println!("Creating a backup of the current GRUB config in '/boot/grub2/grub.cfg.backup'...");
+
+            let current_config_path = grub_config_dir.clone();
+            current_config_path.push(path);
+            fs::copy(current_config_path, "/boot/grub2/grub.cfg.backup").ok_or_else(|| anyhow!("Could not copy the current GRUB config. Not performing migration."));
+    
+            // Copy it again alongside the current symlink
+            fs::copy(current_config_path, "/boot/grub2/grub.cfg.current").ok_or_else(|| anyhow!("Could not copy the current GRUB config. Not performing migration."));
+
+            // Atomically exchange the configs
+            dirfd.local_exchange("grub.cfg.current", "/grub.cfg")?;
+
+            // Remove the now unsused symlink (optional cleanup, ignore any failures)
+            dirfd.remove_file("grub.cfg.current");
+
+            println!("GRUB config symlink successfully replaced with the current config.");
+        },
+    };
+
+    // # If /etc/default/grub exists then we have to force the regeneration of the
+    // # GRUB config to remove the ostree entries that duplicates the BLS ones
+    // if [[ -f "/etc/default/grub" ]]; then
+    //     echo "Marking bootloader as BLS capable..."
+    //     touch "/boot/grub2/.grub2-blscfg-supported"
+
+    //     echo "Regenerating a pure BLS GRUB config..."
+    //     grub2-mkconfig -o /boot/grub2/grub.cfg
+    // fi
+    if !destdir.exists(&first_dir_tmp)? {
+        copy_dir(destdir, first_dir.as_str(), &first_dir_tmp)?;
+        updates.insert(first_dir, first_dir_tmp);
+    }
+
+    // # Remount /sysroot read write just for this unit (we are called in a slave mount namespace by systemd)
+    ensure_writable_sysroot();
```

Comment:
Hm, is this necessary? We're just forking off ostree which knows how to do this.

Reply:
Indeed